IBC.md

```diff
@@ -305,10 +308,157 @@ pub struct IbcPacket {
-https://github.com/CosmWasm/cosmwasm/issues/762)
+##### Acknowledging Errors
+
+A major issue that is unique to `ibc_packet_receive` is that it is expected to
+often reject an incoming packet, yet it cannot abort the transaction. We
+actually expect all state changes from the contract (as well as dispatched
+messages) to be reverted when the packet is rejected, but the transaction to
+properly commit an acknowledgement with encoded error. In other words, this "IBC
+Handler" will error and revert, but the "IBC Router" must succeed and commit an
+acknowledgement message (that can be parsed by the sending chain as an error).
+
+The atomicity issue was first
+[analyzed in the Cosmos SDK implementation](https://github.com/cosmos/ibc-go/issues/68)
+and refined into
+[changing semantics of the OnRecvPacket SDK method](https://github.com/cosmos/ibc-go/issues/91),
+which was
+[implemented in April 2021](https://github.com/cosmos/ibc-go/pull/107), likely
+to be released with Cosmos SDK 0.43 or 0.44. Since we want the best,
+future-proof interface for contracts, we will use an approach inspired by that
+work, and add an adapter in `wasmd` until we can upgrade to a Cosmos SDK version
+that implements this.
+
+After quite some
+[discussion on how to encode the errors](https://github.com/CosmWasm/cosmwasm/issues/762),
+we struggled to map this idea to the CosmWasm model. However, we also discovered
+a deep similarity between these requirements and the
+[submessage semantics](./SEMANTICS.md#submessages). It just requires some
+careful coding on the contract developer's side to new throw errors. This
+produced 3 suggests on how to handle errors and rollbacks _inside
+`ibc_packet_receive`_
+
+1. If the message doesn't modify any state directly, you can simply put the
+   logic in a closure, and capture errors, converting them into error
+   acknowledgements. This would look something like the
+   [main dispatch loop in `ibc-reflect`](https://github.com/CosmWasm/cosmwasm/blob/cd784cd1148ee395574f3e564f102d0d7b5adcc3/contracts/ibc-reflect/src/contract.rs#L217-L248):
+
+```rust
+    (|| {
+        // which local channel did this packet come on
+        let caller = packet.dest.channel_id;
+        let msg: PacketMsg = from_slice(&packet.data)?;
+        match msg {
+            PacketMsg::Dispatch { msgs } => receive_dispatch(deps, caller, msgs),
+            PacketMsg::WhoAmI {} => receive_who_am_i(deps, caller),
+            PacketMsg::Balances {} => receive_balances(deps, caller),
+        }
+    })()
+    .or_else(|e| {
+        // we try to capture all app-level errors and convert them into
+        // acknowledgement packets that contain an error code.
+        let acknowledgement = encode_ibc_error(format!("invalid packet: {}", e));
+        Ok(IbcReceiveResponse {
+            acknowledgement,
+            submessages: vec![],
+            messages: vec![],
+            attributes: vec![],
+        })
+    })
+```
```

Comment:
We could even enforece this by changing the signature of the entry point to `IbcReceiveResponse`, i.e. no result around it.

Reply:
Nice idea. This will push devs to proper usage